src/prometheus/ser.rs

```diff
@@ -0,0 +1,406 @@
+
+#[derive(Debug)]
+pub struct CustomLabelKey {
+    pub name: String,
+    pub help: String,
+    pub key: String,
+}
+
+#[derive(Debug)]
+struct GaugeState {
+    label_key: String,
+    primary_value: String,
+    secondary_value: String,
+    gauge_primary_opt: Opts,
+    gauge_secondary_opt: Opts,
+    name: String,
+}
+pub fn to_prom<T>(
+    value: &T,
+    include: Vec<String>,
+    primary: CustomLabelKey,
+    secondary: CustomLabelKey,
+    label_type: String,
+) -> Result<String>
+where
+    T: Serialize,
+{
+    let registry = Registry::new();
+    let filter_label = FilterLabel {
+        include,
+        primary,
+        secondary,
+        label_type,
+    };
+    let mut serializer = Serializer {
+        gauges: HashMap::new(),
+        gauge_state: None,
+        filter_label,
+    };
+    value.serialize(&mut serializer)?;
+
+    for gauge_list in serializer.gauges {
+        for gauge in gauge_list.1 {
+            registry.register(Box::new(gauge)).unwrap_or_default();
+        }
+    }
+    let mut buffer = Vec::<u8>::new();
+    let encoder = TextEncoder::new();
+    let metric_families = registry.gather();
+    let Ok(_) = encoder.encode(&metric_families, &mut buffer)  else {
+        return Err(Error::Message("Can not encode ".to_string()))
+    };
+    Ok(String::from_utf8(buffer.clone()).unwrap_or_default())
+}
+
+impl Serializer {
+    fn serialize_value<V>(&mut self, v: V) -> Result<()>
+    where
+        V: ToString,
+    {
+        let Some(gauge_state) = &mut self.gauge_state else {
+            return Err(Error::ExpectedStartStruct);
+        };
+        let primary_key = self.filter_label.primary.key.to_string();
+        let secondary_key = self.filter_label.secondary.key.to_string();
+        let label_type = self.filter_label.label_type.to_string();
+        let label_key = &gauge_state.label_key;
+        match label_key {
+            label_key if *label_key != primary_key && *label_key != secondary_key => {
+                let gauge_primary_opt = gauge_state
+                    .gauge_primary_opt
+                    .clone()
+                    .const_label(gauge_state.label_key.clone(), v.to_string());
+
+                let gauge_secondary_opt = gauge_state
+                    .gauge_secondary_opt
+                    .clone()
+                    .const_label(gauge_state.label_key.clone(), v.to_string());
+
+                gauge_state.gauge_primary_opt = gauge_primary_opt;
+                gauge_state.gauge_secondary_opt = gauge_secondary_opt;
+            }
+            &_ => (),
+        }
+        match label_key {
+            label_key if *label_key == primary_key => gauge_state.primary_value = v.to_string(),
+            label_key if *label_key == secondary_key => gauge_state.secondary_value = v.to_string(),
+            label_key if *label_key == label_type => {
+                gauge_state.name = gauge_state.name.clone() + &v.to_string()
+            }
+            &_ => (),
+        }
```

Comment:
Is there a specific reason not to merge theses matches ?
```rust
match label_key {
            label_key if *label_key == primary_key => gauge_state.primary_value = v.to_string(),
            label_key if *label_key == secondary_key => gauge_state.secondary_value = v.to_string(),
            _ => {
                if label_key == label_type {
                   gauge_state.name = gauge_state.name.clone() + &v.to_string()
                }
                ...
}
```

Reply:
Done